Userland/Libraries/LibGfx/ImageFormats/TGALoader.cpp

```diff
@@ -174,55 +62,84 @@ IntSize TGAImageDecoderPlugin::size()
-
-    m_context->header.id_length = reader->read_u8();
-    m_context->header.color_map_type = reader->read_u8();
-    m_context->header.data_type_code = static_cast<TGADataType>(reader->read_u8());
-    m_context->header.color_map_origin = reader->read_i16();
-    m_context->header.color_map_length = reader->read_i16();
-    m_context->header.color_map_depth = reader->read_u8();
-    m_context->header.x_origin = reader->read_i16();
-    m_context->header.y_origin = reader->read_i16();
-    m_context->header.width = reader->read_u16();
-    m_context->header.height = reader->read_u16();
-    m_context->header.bits_per_pixel = reader->read_u8();
-    m_context->header.image_descriptor = reader->read_u8();
-
-    auto bytes_remaining = reader->data().size() - reader->index();
-
-    if (m_context->header.data_type_code == TGADataType::UncompressedRGB && bytes_remaining < static_cast<u64>(m_context->header.width) * m_context->header.height * (m_context->header.bits_per_pixel / 8))
-        return Error::from_string_literal("Not enough data to read an image with the expected size");
-
-    if (m_context->header.bits_per_pixel < 8 || m_context->header.bits_per_pixel > 32)
+    auto bytes_remaining = whole_image_stream_size - sizeof(TGAHeader);
+    if ((header.bits_per_pixel % 8) != 0 || header.bits_per_pixel < 8 || header.bits_per_pixel > 32)
         return Error::from_string_literal("Invalid bit depth");
+    if (header.data_type_code == TGADataType::UncompressedRGB && bytes_remaining < static_cast<u64>(header.width) * header.height * (header.bits_per_pixel / 8))
+        return Error::from_string_literal("Not enough data to read an image with the expected size");
+    return {};
+}
 
+ErrorOr<void> TGAImageDecoderPlugin::decode_tga_header()
+{
+    // NOTE: Read the image size before reading the header from the stream
+    // as it will truncate that from what we need - the whole image size, including
+    // the header size in it.
+    auto image_bytes_size = static_cast<FixedMemoryStream const&>(m_context->stream).bytes().size();
+    TRY(m_context->stream.read_until_filled(Bytes { reinterpret_cast<u8*>(&m_context->header), sizeof(TGAHeader) }));
+    TRY(ensure_header_validity(m_context->header, image_bytes_size));
     return {};
 }
 
 ErrorOr<bool> TGAImageDecoderPlugin::validate_before_create(ReadonlyBytes data)
 {
-    if (data.size() < sizeof(TGAHeader))
-        return false;
-    TGAHeader const& header = *reinterpret_cast<TGAHeader const*>(data.data());
-    if (header.data_type_code == TGADataType::UncompressedRGB && data.size() < static_cast<u64>(header.width) * header.height * (header.bits_per_pixel / 8))
-        return false;
-    if (header.bits_per_pixel < 8 || header.bits_per_pixel > 32)
-        return false;
-    return true;
+    FixedMemoryStream stream { data };
+    TGAHeader header {};
+    TRY(stream.read_until_filled(Bytes { reinterpret_cast<u8*>(&header), sizeof(TGAHeader) }));
```

Comment:
If you want, you can register `TGAHeader` as being `is_trivially_serializable`, so you will be able to call `read_value<TGAHeader>()`